File: CODE/funRastrigin.f90
!===============================================================
!
! Funzione Rastrigin (n variabile) consigliata n=5
! ALI-TORN POPULATION SET BASED GO ALGORITHMS: SOME MODIFICATIONS AND NUMERICAL STUDIES
! COMPUTERS AND OPERATION RESEARCH 
!===============================================================
!			VARIABLE BOUNDS
!===============================================================
!       -5.12D0              5.12D0          Limits of variable i
! FGLOB = 0.0D0 XSTAR = 0.0D0

subroutine funct(n,x,f)

      implicit none

      integer          :: n
      double precision :: x(n), f, xaux(n)

      double precision, parameter :: pi = 3.1415923d0

      integer          :: i

	  xaux = x + 2.56d0*dsqrt(2.d0)

      f = 10.d0*dble(n)

	  do i = 1,n
		f = f + xaux(i)**2.d0 -10.d0*dcos(2.d0*pi*xaux(i))
	  enddo

      return

end


subroutine grad(n,x,g)

implicit none

integer			::n, i
double precision::x(n), g(n), xaux(n)
double precision, parameter :: pi = 3.1415923d0

xaux = x + 2.56d0*dsqrt(2.d0)

do i=1,n

	g(i) = 2.0d0*xaux(i)+20.0d0*pi*dsin(2.d0*pi*xaux(i))

enddo

return

end subroutine grad

File: Python/funRastrigin.py
#===============================================================
#
# Funzione Rastrigin (n variabile) consigliata n=5
# ALI-TORN POPULATION SET BASED GO ALGORITHMS: SOME MODIFICATIONS AND NUMERICAL STUDIES
# COMPUTERS AND OPERATION RESEARCH 
#===============================================================
#			VARIABLE BOUNDS
#===============================================================
#       -5.12D0              5.12D0          Limits of variable i
# FGLOB = 0.0D0 XSTAR = 0.0D0

import numpy as np

def rastrigin_function(x):
    # Function to evaluate the Rastrigin function
    n = len(x)
    x_aux = x + 2.56 * np.sqrt(2.0)
    f = 10.0 * n

    for i in range(n):
        f += x_aux[i]**2 - 10.0 * np.cos(2.0 * np.pi * x_aux[i])

    return f

def rastrigin_gradient(x):
    # Function to compute the gradient of the Rastrigin function
    n = len(x)
    x_aux = x + 2.56 * np.sqrt(2.0)
    g = np.zeros_like(x)

    for i in range(n):
        g[i] = 2.0 * x_aux[i] + 20.0 * np.pi * np.sin(2.0 * np.pi * x_aux[i])

    return g

if __name__ == "__main__":
    # Example usage
    n_dim = 5
    x_example = np.zeros(n_dim)

    function_value = rastrigin_function(x_example)
    gradient_value = rastrigin_gradient(x_example)

    print("Function value:", function_value)
    print("Gradient value:", gradient_value)
